### PR TITLE
NXP-27483: implement clid endpoint

### PR DESCRIPTION
**Pull request by NourNuxeo**

### CODE REVIEW

File: nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/clid/ClidObject.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ * @since 11.1
+ */
+@WebObject(type = "clid")
+public class ClidObject extends DefaultObject {
+
+    @GET
+    @Produces(MediaType.APPLICATION_JSON)
+    public Object doGet() throws NoCLID {
+        String clid = LogicalInstanceIdentifier.instance().getCLID();
+        Map<String, Serializable> map = new HashMap<>();
```
Comment (ataillefer) [inline nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/clid/ClidObject.java]: `Map<String,String>` is enough

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    @Test
+    public void testClid() throws IOException, NoCLID, InvalidCLID {
+        String expected = "choco--stick";
+        new LogicalInstanceIdentifier(expected).save();
+        try (CloseableClientResponse response = get()) {
+            assertEquals(200, response.getStatus());
+            JsonNode node = mapper.readTree(response.getEntityInputStream());
+            JsonNode key = node.get("clid");
```
Comment (ataillefer) [inline nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java]: "key" doesn't seem appropriate for a value

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    protected ObjectMapper mapper;
+
+    @Before
+    public void setup() {
+        mapper = new ObjectMapper();
+        client = JerseyClientHelper.clientBuilder().setCredentials("Administrator", "Administrator").build();
+    }
+
+    @Test
+    public void testClid() throws IOException, NoCLID, InvalidCLID {
```
Comment (ataillefer) [inline nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java]: Please remove `NoCLID`, Sonar should warn about it.

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        String expected = "choco--stick";
+        new LogicalInstanceIdentifier(expected).save();
+        try (CloseableClientResponse response = get()) {
+            assertEquals(200, response.getStatus());
+            JsonNode node = mapper.readTree(response.getEntityInputStream());
+            JsonNode key = node.get("clid");
+            String clid = LogicalInstanceIdentifier.instance().getCLID();
+            assertEquals(clid, key.asText());
```
Comment (ataillefer) [inline nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java]: I think these 2 lines are superfluous regarding the line below.

**Comment by ataillefer** [comment]
Can you please add a test with an invalid CLID? `NoCLID` can be thrown, which should result in a 500 error.

File: nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/clid/ClidObject.java (hunk 1 above)
Comment (RSalem07) [reply nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/clid/ClidObject.java]: regarding to [`NXP-27483`](https://jira.nuxeo.com/browse/NXP-27483) this feature is only for master (Java 11). If it's the case then it can be replaced by `Map.of` 
`Map<String, String> clid = Map.of("clid", LogicalInstanceIdentifier.instance().getCLID());` 
wdyt :)

Comment (RSalem07) [reply nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/clid/ClidObject.java]: is there a raison to use Object as the returned type of this method. i think `Response` can be used wdyt 
@ataillefer any adivce.

Comment (RSalem07) [reply nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/clid/ClidObject.java]: i am wonedring about what if the reste resource use try/catch bloc and then return an 500 / 404 error + Log in the case of the exception NoCLID. i don't know if is there any servlet filter that manage exception in api rest management like the case of Nuxeo that holds exceptions and then log / send the http status @ataillefer wdyt

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java (hunk 3 above)
Comment (RSalem07) [reply nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java]: for the use of `mapper` and `client` i think the code can be refactored with BaseTest class to hold these 2 fields. the classe `TestDistributionObject` already use them with the same values.

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        int port = servletContainerFeature.getPort();
+        return "http://localhost:" + port + "/";
+    }
+
+    protected CloseableClientResponse get() {
+        WebResource wr = client.resource(getBaseURL() + CLID_PATH);
+        WebResource.Builder builder = wr.getRequestBuilder();
+        return CloseableClientResponse.of(builder.get(ClientResponse.class));
+    }
+}
```
Comment (RSalem07) [inline nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java:61]: same as above on the creation of the base test  to add  `getBaseURL` and `get` method which can take a to avoid duplicated code.

File: nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/clid/ClidObject.java (hunk 1 above)
Comment (ataillefer) [reply nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/clid/ClidObject.java]: For the fixVersion, sorry it's a mistake (fixed), for now it's only for 10.10, so Java8.

Comment (ataillefer) [reply nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/clid/ClidObject.java]: Use `Response` yes.

Comment (ataillefer) [reply nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/clid/ClidObject.java]: About handling the exception, yes that's why a test needs to be added.
Probably need to add the `WebEngineExceptionMapper` to the `ManagementModule`'s singletons to handle the exception. Then in this case I don't think try/catch is necessary since the default error code will be 500 which is fine here from a client's point of view.

The `WebEngineExceptionMapper` will log the 500 error, see https://github.com/nuxeo/nuxeo/blob/master/nuxeo-webengine/nuxeo-webengine-core/src/main/java/org/nuxeo/ecm/webengine/app/WebEngineExceptionMapper.java#L80.

Ultimately, a custom exception mapper for the management REST API could be used, as we did for WOPI: https://github.com/nuxeo/nuxeo/blob/master/addons/nuxeo-wopi/nuxeo-wopi/src/main/java/org/nuxeo/wopi/jaxrs/WOPIExceptionMapper.java

But I guess that for now the webengine one should do.

Comment (RSalem07) [reply nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/clid/ClidObject.java]: top, it's what i was looking for. i agree with the wrapper no need for try / catch bloc, the  exception mapper will be reponsible of that.
thanks

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java (hunk 3 above)
Comment (ataillefer) [reply nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java]: Nice idea but:
- `BaseTest` is not exposed since the `nuxeo-rest-api-test` module is not built as a test-jar (could do it but not sure we want to for 10.10 and would add a dependency).
- Not sur if we want to reuse it anyway as it's old and more "document/Nuxeo REST API" oriented.
- In the future, authentication might change: using a JWT token, so building our own client is fine.

If we want to refactor I'm in favor of:
- See if we can use `HttpClientTestRule`, accessible via `nuxeo-features-test` and lighter.
- Otherwise have an abstract test class similar as `BaseTest` but with the strict minimum.

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java (hunk 5 above)
Comment (ataillefer) [reply nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java:61]: See https://github.com/nuxeo/nuxeo-management-rest-api/pull/3#discussion_r293355229

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java (hunk 3 above)
Comment (RSalem07) [reply nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java]: for `BaseTest` i was thinking about a new one, yep i see you point of view. try the two propositions ( `HttpClientTestRule ` and a t least make a new class with the strict minimum it's a good alternative) 
thx

File: nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/clid/ClidObject.java (hunk 1 above)
Comment (NourNuxeo) [reply nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/clid/ClidObject.java]: NoClid => 500 💪

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java (hunk 3 above)
Comment (NourNuxeo) [reply nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java]: will do in a separate commit then

File: nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/clid/ClidObject.java (hunk 1 above)
Comment (ataillefer) [reply nuxeo-management-rest-api/src/main/java/org/nuxeo/rest/management/clid/ClidObject.java]: OK, you get a 500 and an ERROR log server-side because jersey's `ContainerResponse` handles it by default, but it's not really nice and what we expect.

You'll probably get something like
```
ERROR [ContainerResponse] The exception contained within MappableContainerException could not be mapped to a response, re-throwing to the HTTP container
```

in the server logs, and a an ugly HTML error page as a response of the curl call.

Using the `WebEngineExceptionMapper`, you should get:
- A clean error logged server-side by the `WebEngineExceptionMapper`
- Client-side, a JSON reponse with the exception message, cause and stack trace (and still status code 500 of course).

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        Builder builder = new HttpClientTestRule.Builder().accept(JSONBlob.APPLICATION_JSON)
+                                                          .credentials(ADMINISTRATOR, ADMINISTRATOR)
+                                                          .url(url);
+        HttpClientTestRule response = builder.build();
+        response.starting();
+        CloseableClientResponse result = response.get(path);
+        response.finished();
+        return result;
+    }
+}
```
Comment (RSalem07) [inline nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java]: thanks, :)
its seem good, but i think we can make some improvements (nice to have) like 
`protected ObjectMapper mapper = new ObjectMapper();
 CloseableClientResponse get(String path) {
        Builder builder = new HttpClientTestRule.Builder().accept(JSONBlob.APPLICATION_JSON)
                .credentials(ADMINISTRATOR, ADMINISTRATOR)
                .url(String.format("http://localhost:%s/", servletContainerFeature.getPort()));

        HttpClientTestRule response = builder.build();
        response.starting();
        CloseableClientResponse result = response.get(path);
        response.finished();
        return result;
    }`
-  I dont think the `before` method is required, we can we create the object mapper on in line way
- we can replace the url var creation by String#format

for the status codes checking, i think use the `org.apache.http.HttpStatus` and 
`assertEquals(HttpStatus.SC_OK, response.getStatus());` with an static import on code 
assertEquals(SC_OK, response.getStatus());

wdyt,

Comment (NourNuxeo) [reply nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java]: The @before executes before each test. That's why i left the setup method.
What do you think about it ?
About status code ok for me !

Comment (RSalem07) [reply nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java]: yep i agree for `@before` method is before each test, but the `ObjectMapper` is thread safe, is just used to read tthe json response, stateless. a unique instance is is fully enough, we can have another point view from @ataillefer

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import org.nuxeo.runtime.test.runner.ServletContainerFeature;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * @since 11.1
+ */
+@RunWith(FeaturesRunner.class)
+@Features(ManagementFeature.class)
+public class BaseTest {
```
Comment (ataillefer) [inline nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java]: Can you please rename it to `ManagementBaseTest` to avoid the confusion with the existing `BaseTest` class?

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java (hunk 6 above)
Comment (ataillefer) [reply nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java]: - Yes about ObjectMapper initialized inline => no `@Before` needed for this.
- Sure for String#format instead of +, though it's really a detail here in the context of unit tests.
Yet, I'm personally still in favor of defining some intermediate variables such as `url` as it's more readable and eases debugging.
I'd write:
```
String url = String.format("http://localhost:%s/", servletContainerFeature.getPort());
call(url)
```

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+import org.nuxeo.runtime.test.runner.Features;
+import org.nuxeo.runtime.test.runner.FeaturesRunner;
+
+import com.fasterxml.jackson.databind.JsonNode;
+
+/**
+ * @since 11.1
+ */
+@RunWith(FeaturesRunner.class)
+@Features(ManagementFeature.class)
```
Comment (ataillefer) [inline nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/TestClidObject.java]: You should be able to remove these 2 lines as they are in the parent class, same for the other test.

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java (hunk 7 above)
Comment (ataillefer) [reply nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java]: You also need to make this class `abstract`, otherwise mvn will complain about "No runnable methods".

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    protected ObjectMapper mapper;
+
+    @Before
+    public void setup() {
+        mapper = new ObjectMapper();
+    }
+
+    protected CloseableClientResponse get(String path) {
+        String url = "http://localhost:" + servletContainerFeature.getPort() + "/";
+        Builder builder = new HttpClientTestRule.Builder().accept(JSONBlob.APPLICATION_JSON)
```
Comment (ataillefer) [inline nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java]: I would use `javax.ws.rs.coreMediaType.APPLICATION_JSON` imported as a static import rather than a custom Nuxeo class that is not related.

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java (hunk 6 above)
Comment (ataillefer) [reply nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java]: About the status codes:
- Usually we use `javax.ws.rs.core.Response.Status` or `javax.servlet.http.HttpServletResponse`, see https://github.com/nuxeo/nuxeo/blob/master/addons/nuxeo-wopi/nuxeo-wopi/src/main/java/org/nuxeo/wopi/jaxrs/FilesEndpoint.java#L23
- Anyway, in practice, I find it clearer to have the explicit integer rather than a constant for the expected value. My opinion...

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java (hunk 7 above)
Comment (RSalem07) [reply nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java]: 👍

File: nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java (hunk 6 above)
Comment (ataillefer) [reply nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java]: In fact, you should really take example on how the `HttpClientTestRule` is used, only one test AFAIK: https://github.com/nuxeo/nuxeo/blob/10.10/nuxeo-features/nuxeo-platform-forms-layout-export/src/test/java/org/nuxeo/ecm/platform/forms/layout/export/TestLayoutResource.java

So you'll need à `@Before` to initialize the rule (protected member) and start it, an `@After` to finish it, and no more need of the `get` method, just call `httpClientRule.get` directly!

Comment (RSalem07) [reply nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java]: agree for using the `HttpServletResponse `. Use of constants, give us a short descritpion of the status, mainly where we expect a status that we don't use often. It's a nice to have and not mandatory. I think the idea is to be consistent with the other code of the project.

Comment (ataillefer) [reply nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java]: Sure, whatever works!

Comment (NourNuxeo) [reply nuxeo-management-rest-api/src/test/java/org/nuxeo/rest/management/BaseTest.java]: Implemented all that. About the status code I used the HttpServletResponse. But I like numbers better. I actually had to look up for which one of the constants would send a 500

**Review by ataillefer — APPROVED** [review]

**Review by RSalem07 — APPROVED** [review]
